prov/hook/profile/src/prof_config.c

```diff
@@ -0,0 +1,494 @@
+		                                  ops, num_msgsize_event,
+		                                  (void *)msgsize_events);
+	}
+
+    return ret;
+}
+
+static int
+prof_config_cq_data_class(struct profile_context *ctx) {
+	int ret = 0;
+	struct profile_cntr_class *cls;
+
+	ret = prof_add_class(ctx, HOOK_PROF_CQDATA, HOOK_PROF_CQDATA_NAME,
+	                     HOOK_PROF_CQ_MAX, 0,
+	                     prof_update_msgsize_counter,
+	                     prof_report_msgsize_counter);
+	if (ret < 0)
+		return ret;
+
+	cls = &(ctx->cntr_classes[HOOK_PROF_CQDATA]);
+	for (int i = 0; i < HOOK_PROF_CQ_MAX; i++) {
+		ctx->cntrs_size += prof_add_group(ctx, cls, i, prof_cq_data_str[i],
+		                                  0, num_msgsize_event,
+		                                  (void *)msgsize_events);
+	}
+
+    return ret;
+}
+
+/*
+ * create a context for profile
+ */
+void
+prof_context_create(struct fi_provider *prov, struct profile_context *ctx)
+{
+	int ret = 0;
+	int index = 0;
+	struct profile_cntr_class *cls;
+	struct profile_cntr_group *grp;
+
+	ctx->prov = prov;
+	ctx->cntrs_size = 0;
+	ctx->num_class = 0;
+
+	prof_config_init(ctx);
+	memset(&ctx->cntr_classes, 0, sizeof(ctx->cntr_classes));
+	ret = prof_config_api_class(ctx);
+	if (!ret) {
+		ret = prof_config_cq_data_class(ctx);
+	}
+	if (ret || !ctx->cntrs_size)  {
+		ctx->num_class = 0;
+		FI_TRACE(prov, FI_LOG_CORE, "Create prof context failed, ret %d\n", ret); 
+		return;
+	}
+	ctx->cntrs_mem = (struct profile_counter *)calloc(ctx->cntrs_size,
+	                 sizeof(struct profile_counter));
+	if (!ctx->cntrs_mem) {
+		ctx->num_class = 0;
+		ctx->cntrs_size = 0;
+		FI_TRACE(prov, FI_LOG_CORE, 
+		         "Create prof context: allocate counter memory failed\n"); 
+		return;
+	}
+	for (int i = 0; i < HOOK_PROF_CLASS_MAX; i++) {
+		cls = &(ctx->cntr_classes[i]);
+		for (int j = 0; j < cls->num_group; j++) {
+			grp = &(cls->cntr_groups[j]);
+			if (grp->num_counter) {
+				grp->counters = &(ctx->cntrs_mem[index]);
+				index += grp->num_counter;
+			}
+		}
+	}
+}
+
+void
+prof_add_cntr_data(struct profile_context *ctx, int class_id,
+                   int group, size_t len)
+{
+	struct profile_cntr_class *cls;
+
+	if ((class_id >= ctx->num_class) ||
+	    (group >= ctx->cntr_classes[class_id].num_group))
+		return;
+
+	cls = &(ctx->cntr_classes[class_id]);
+	cls->update_counter(ctx, cls, &(cls->cntr_groups[group]), 0, len);
+}
+
+
+void
+prof_add_cntr_cq_data(struct profile_context *ctx,  int class_id,
+                      enum fi_cq_format format, void *buf, int ret)
+{
+	uint64_t len = 0;
+	int group = 0;
+	struct profile_cntr_class *cls;
+
+	if (class_id >= ctx->num_class)
+		return;
+
+	cls = &(ctx->cntr_classes[class_id]);
+
+	for (int i = 0; i < ret; i++) {
+		if (get_cq_entry[format](buf, i, &group, &len) &&
+		    (group < cls->num_group))  {
+			cls->update_counter(ctx, cls, &cls->cntr_groups[group], 0, len);
+		}
+	}
+}
```

Comment:
I think if we dropped the ability to try to create counters dynamically and use the fixed bucket sizes, we should be able to allocate the counters that we need statically.